url8d1df200a8_gilleain_signatures_tgz-pJ8-signature_chemistry_LargeMoleculeTestJ8/src/signature/chemistry/MoleculeReader.java

```diff
@@ -0,0 +1,135 @@
+package signature.chemistry;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.List;
+
+import signature.chemistry.Molecule.BondOrder;
+
+public class MoleculeReader {
+
+    public static Molecule readMolfile(String filename) {
+        Molecule molecule = null;
+        try {
+            File file = new File(filename);
+            BufferedReader reader = new BufferedReader(new FileReader(file));
+            String line;
+            List<String> block = new ArrayList<String>();
+            while ((line = reader.readLine()) != null) {
+                block.add(line);
+            }
+            molecule = makeMolecule(block);
+            reader.close();
+        } catch (IOException ioe) {
+            System.err.println(ioe.toString());
+        } catch (Exception e) {
+            System.err.println(e.toString());
+        }
+        return molecule;
+    }
+    
+    
+	/**
+	 * Read a list of Molecule from SDFile
+	 * @param filename path to SDFile
+	 * @return List of Molecules
+	 */
+    public static List<Molecule> readSDFFile(String filename) {
+
+        File file = new File(filename);
+        try {
+            return readSDFfromStream(new FileInputStream(file));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor: *FileInputStream*. Execution paths that do not contain closure statements were detected. Ensure the object returned by *FileInputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).